perception/autoware_image_projection_based_fusion/src/camera_projection.cpp

```diff
@@ -0,0 +1,264 @@
+
+  // check each side of the view
+  if (ray_bottom_left.x < x_left.x) x_left = ray_bottom_left;
+  if (ray_mid_left.x < x_left.x) x_left = ray_mid_left;
+
+  if (x_right.x < ray_bottom_right.x) x_right = ray_bottom_right;
+  if (x_right.x < ray_mid_right.x) x_right = ray_mid_right;
+
+  if (y_top.y < ray_top_right.y) y_top = ray_top_right;
+  if (y_top.y < ray_mid_top.y) y_top = ray_mid_top;
+
+  if (ray_bottom_left.y < y_bottom.y) y_bottom = ray_bottom_left;
+  if (ray_mid_bottom.y < y_bottom.y) y_bottom = ray_mid_bottom;
+
+  // set FOV at z = 1.0
+  fov_left_ = x_left.x / x_left.z;
+  fov_right_ = x_right.x / x_right.z;
+  fov_top_ = y_top.y / y_top.z;
+  fov_bottom_ = y_bottom.y / y_bottom.z;
+}
+
+void CameraProjection::initialize()
+{
+  if (unrectify_) {
+    if (use_approximation_) {
+      // create the cache with size of grid center
+      // store only xy position in float to reduce memory consumption
+      projection_cache_ = std::make_unique<PixelPos[]>(cache_size_);
+      initializeCache();
+
+      calcImageProjectedPoint = [this](
+                                  const cv::Point3d & point3d, Eigen::Vector2d & projected_point) {
+        return this->calcRawImageProjectedPointWithApproximation(point3d, projected_point);
+      };
+    } else {
+      calcImageProjectedPoint = [this](
+                                  const cv::Point3d & point3d, Eigen::Vector2d & projected_point) {
+        return this->calcRawImageProjectedPoint(point3d, projected_point);
+      };
+    }
+  } else {
+    calcImageProjectedPoint = [this](
+                                const cv::Point3d & point3d, Eigen::Vector2d & projected_point) {
+      return this->calcRectifiedImageProjectedPoint(point3d, projected_point);
+    };
+  }
+}
+
+void CameraProjection::initializeCache()
+{
+  // sample grid centers till the camera height, width to precompute the projection
+  //
+  //      grid_size
+  //      /
+  //     v
+  //   |---|          w
+  //  0----------------->
+  // 0 | . | . | . |
+  //   |___|___|___|
+  //   | . | . | . |
+  //   | ^
+  // h | |
+  //   v grid center
+  //
+  // each pixel will be rounded in this grid center
+  // edge pixels in the image will be assign to centers that is the outside of the image
+
+  for (float y = half_grid_h_size_; y < image_h_; y += grid_h_size_) {
+    for (float x = half_grid_w_size_; x < image_w_; x += grid_w_size_) {
+      const float qx =
+        std::round((x - half_grid_w_size_) * inv_grid_w_size_) * grid_w_size_ + half_grid_w_size_;
+      const float qy =
+        std::round((y - half_grid_h_size_) * inv_grid_h_size_) * grid_h_size_ + half_grid_h_size_;
+
+      const int grid_x = static_cast<int>(std::floor(qx / grid_w_size_));
+      const int grid_y = static_cast<int>(std::floor(qy / grid_h_size_));
+      const int index = (grid_y)*grid_x_num_ + grid_x;
+
+      // precompute projected point
+      cv::Point2d raw_image_point = camera_model_.unrectifyPoint(cv::Point2d(qx, qy));
```

Comment:
```suggestion
  for (int grid_x = 0; y < image_h_/grid_h_size_; y++) {
    for (int grid_x = 0; x < image_w_ /grid_w_size_; x++) {
      const int index = (grid_y)*grid_x_num_ + grid_x;
      const float qx = (grid_x + 0.5f) * grid_w_size_;
      const float qy = (grid_y + 0.5f) * grid_h_size_;
      
      // precompute projected point
      cv::Point2d raw_image_point = camera_model_.unrectifyPoint(cv::Point2d(qx, qy));
```
Better readability and efficiency.
(need to be tested)

Reply:
Thank you for the suggestion!
As same as the above suggestion, it is added in https://github.com/autowarefoundation/autoware.universe/pull/9635/commits/e8b5b2ffdcfcca886f51c9d58f26f623356a5932.